internal/docker/client.go

```diff
@@ -384,41 +386,110 @@ func (c *Cli) ServerVersion() types.Version {
+func (c *Cli) authInfo(ctx context.Context, repoInfo *registry.RepositoryInfo, cmdName string) (encodedAuth, types.RequestPrivilegeFunc, error) {
 	infoWriter := logger.Get(ctx).Writer(logger.InfoLvl)
 	cli, err := command.NewDockerCli(
 		command.WithCombinedStreams(infoWriter),
 		command.WithContentTrust(true),
 	)
 	if err != nil {
-		return nil, errors.Wrap(err, "ImagePush#NewDockerCli")
+		return "", nil, errors.Wrap(err, "authInfo#NewDockerCli")
 	}
 
 	err = cli.Initialize(cliflags.NewClientOptions())
 	if err != nil {
-		return nil, errors.Wrap(err, "ImagePush#InitializeCLI")
+		return "", nil, errors.Wrap(err, "authInfo#InitializeCLI")
 	}
 	authConfig := command.ResolveAuthConfig(ctx, cli, repoInfo.Index)
-	requestPrivilege := command.RegistryAuthenticationPrivilegedFunc(cli, repoInfo.Index, "push")
+	requestPrivilege := command.RegistryAuthenticationPrivilegedFunc(cli, repoInfo.Index, cmdName)
+
+	auth, err := command.EncodeAuthToBase64(authConfig)
+	if err != nil {
+		return "", nil, errors.Wrap(err, "authInfo#EncodeAuthToBase64")
+	}
+	return encodedAuth(auth), requestPrivilege, nil
+}
 
-	encodedAuth, err := command.EncodeAuthToBase64(authConfig)
+func (c *Cli) ImagePull(ctx context.Context, ref reference.Named) (reference.Canonical, error) {
+	repoInfo, err := registry.ParseRepositoryInfo(ref)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse registry for %q: %v", ref.String(), err)
+	}
+
+	encodedAuth, requestPrivilege, err := c.authInfo(ctx, repoInfo, "push")
+	if err != nil {
+		return nil, fmt.Errorf("could not authenticate: %v", err)
+	}
+
+	image := ref.String()
+	pullResp, err := c.Client.ImagePull(ctx, image, types.ImagePullOptions{
+		RegistryAuth:  string(encodedAuth),
+		PrivilegeFunc: requestPrivilege,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not pull image %q: %v", image, err)
+	}
+	defer func() {
+		_ = pullResp.Close()
+	}()
+
+	// the /images/create API is a bit chaotic, returning JSON lines of status as it pulls
+	// including ASCII progress bar animation etc.
+	// there's not really any guarantees with it, so the prevailing guidance is to try and
+	// inspect the image immediately afterwards to ensure it was pulled successfully
+	// (this is racy and could be improved by _trying_ to get the digest out of this response
+	// and making sure it matches with the result of inspect, but Docker itself suffers from
+	// this same race condition during a docker run that triggers a pull, so it's reasonable
+	// to deem it as acceptable here as well)
+	_, err = io.Copy(io.Discard, pullResp)
+	if err != nil {
+		return nil, fmt.Errorf("connection error while pulling image %q: %v", image, err)
+	}
+
+	imgInspect, _, err := c.ImageInspectWithRaw(ctx, image)
+	if err != nil {
+		return nil, fmt.Errorf("failed to inspect after pull for image %q: %v", image, err)
+	}
+
+	digestInfo := strings.SplitN(imgInspect.RepoDigests[0], "@", 2)
```

Comment:
can you add some comments on why you need to hand-write a digest parser here? I thought the docker/distribution parser takes care of this.

Reply:
Sortaaa, `reference.Parse(imgInspect.RepoDigests[0])` + cast to `reference.Canonical` _will_ work, but you'll lose repo (if it's `library.docker.io`) and tag (which is arguably redundant, but useful in this context if you plan to immediately use the value to create+run containers so that it shows up in Docker engine).

(e.g. with the impl here you get `docker.io/nginx:1.21.3@sha256:<hash>` vs just `nginx@sha256:<hash>`)

If you think that's preferable, I can swap over. Unfortunately, `docker/distribution` doesn't have better helpers here (unlike containerd: https://github.com/containerd/containerd/blob/6416cc35c1f0a92a3561411de4ccfb71df2c8308/reference/reference.go#L155-L166)

Reply:
does ParseAnyReference work?

The docker.io/nginx vs nginx is called "Familiarization" in the API, and you can usually handle it with https://pkg.go.dev/github.com/docker/distribution@v2.7.1+incompatible/reference#FamiliarString)

and ya, the image reference spec is a mess, i've been meaning to find one of the authors at kubecon sometime and pry the story out of them of why it's such a mess.

Reply:
Okay, tweaked this - it uses `ParseNormalizedNamed` (`ParseAnyReference` would delegate to this in this case so saves extra cast/indirection), casts to `reference.Canonical` and then uses `reference.WithDigest(original, inspectResult.Digest()`.

Roundabout but best that can be done afaict with their APIs that will ensure:
 * Qualified name
 * Tag preserved if provided to pull
 * Digest attached